tests/Unit/Framework/Tests/Test_Pypp.cpp

```diff
@@ -411,6 +412,35 @@ void test_function_of_time() {
   check(x_d, t);
   check(x_dv, t);
 }
+
+template <template <class> class Optional>
+void test_optional() {
+  const Optional<Scalar<double>> scalar_double_a{Scalar<double>{0.8}};
+  const Optional<Scalar<double>> scalar_double_b{Scalar<double>{1.8}};
+
+  const Optional<Scalar<DataVector>> scalar_datavector_a{
+      Scalar<DataVector>{{{{0.8, 0.7, 0.5}}}}};
+  const Optional<Scalar<DataVector>> scalar_datavector_b{
+      Scalar<DataVector>{{{{1.8, 2.3, 4.2}}}}};
+
+  const auto impl = [](const auto& a, const auto& b) {
+    using T = typename std::decay_t<decltype(a)>::value_type;
+    const auto result_double_double =
+        pypp::call<T>("PyppPyTests", "add_scalars", a, b);
+    CHECK_ITERABLE_APPROX(get(result_double_double), get(T{get(*a) + get(*b)}));
+
+    const auto result_double_none =
+        pypp::call<T>("PyppPyTests", "add_scalars", a, Optional<T>{});
+    CHECK_ITERABLE_APPROX(get(result_double_none), get(*a));
+
+    const auto result_none_double =
```

Comment:
the `none_none` case appears to be in the pypp py part, but is not tested in the test.

Reply:
Good catch! This would require adding of `make_with_value`s for `*::optional<T>` and also handling `optional<T>` at the `ToPyObject` level and `FromPyObject` level. I'm going to assume we aren't going to be returning a `None` from the python code, though support I think could be added in the future. Part of the issue is that we choose the `CallImpl` implicitly based on the return type, so returning `optional<T>` is not so easy right now.  Probably the best way to handle this would be to have a `call_over_grid` function in addition to the `call` function already in `pypp`. We then explicitly know whether we need to loop or not. I'd prefer to defer that level of change, though since that'll propagate to more than 100 targets/files.